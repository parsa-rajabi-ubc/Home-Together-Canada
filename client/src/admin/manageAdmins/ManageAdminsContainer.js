/**
 *
 * @Author:     Parsa Rajabi
 * @Created:    2021.3.12
 *
 * @Description: Manage Admins
 *
 */

import React, {useEffect, useState} from "react";
import ManageAdmins from "./ManageAdmins";
import * as AdminService from "../../services/AdminService";
import * as MemberService from '../../services/MemberService';
import {
    checkIfErrorsExistInMapping,
    getConcatenatedErrorMessage,
    validateInput
} from "../../registration/registrationUtils";
import {Flip, toast} from "react-toastify";
import 'react-toastify/dist/ReactToastify.css'
import {ADMIN_TOAST} from "../../common/constants/ToastText";


toast.configure()

function ManageAdminsContainer() {
    const [searchUsername, setSearchUsername] = useState();
    const [searchUsernameError, setSearchUsernameError] = useState();
    const [currentAdmins, setCurrentAdmins] = useState([]);


    // useEffects
    useEffect(() => {
        searchUsername !== undefined && validateInput(searchUsername, setSearchUsernameError);
    }, [searchUsername]);

    useEffect(() => {
        getAllAdmins();
    }, [])


    const getAllAdmins = () => {
        AdminService.getAllAdmins()
            .then(res => res.json())
            .then(data => {
                setCurrentAdmins(data.admins);
            })
    }
    const isSearchValid = () => {
        const searchErrors = {
            username: false,
        }
        searchErrors.username = validateInput(searchUsername, setSearchUsernameError);

        // check search criteria for errors
        return !checkIfErrorsExistInMapping(searchErrors);
    }
    const onSearch = () => {
        const promoteMemberBodyRequest = {
            username: searchUsername
        }
        if (isSearchValid()) {

            MemberService.findMemberProfileByUsername(searchUsername)
                .then(res => res.json())
                .then(data => {
                    if (data.profile) {
                        AdminService.promoteMemberToAdmin(promoteMemberBodyRequest)
                            .then(res => res.json())
                            .then(data => {
                                if (data.adminPrivileges) {
                                    toast.success(ADMIN_TOAST.MANAGE_ADMIN_SUCCESS, {
                                        toastId: "successToast",
                                        position: "bottom-center",
                                        autoClose: 10000,
                                        hideProgressBar: false,
                                        closeOnClick: true,
                                        pauseOnHover: true,
                                        draggable: true,
                                        progress: false,
                                        transition: Flip
                                    });
                                    getAllAdmins();

                                } else if (data.err) {
                                    alert('Error: ' + data.err);

                                } else if (data.errors) {
                                    const errorMessage = getConcatenatedErrorMessage(data.errors);
                                    // show list of all errors
                                    alert(errorMessage);
                                }
                            })
                            .catch(err => {
                                alert('Error: ' + err.message);
                            });
                    } else {
                        toast.error(ADMIN_TOAST.MANAGE_ADMIN_USERNAME_DNE, {
                            toastId: "usernameDNE",
                            position: "bottom-center",
                            autoClose: 10000,
                            hideProgressBar: false,
                            closeOnClick: true,
                            pauseOnHover: true,
                            draggable: true,
                            progress: false,
                            transition: Flip
                        });
                    }
                });
        } else {
            toast.error(ADMIN_TOAST.MANAGE_ADMIN_EMPTY_USERNAME, {
                toastId: "emptyUsername",
                position: "bottom-center",
                autoClose: 10000,
                hideProgressBar: false,
                closeOnClick: true,
                pauseOnHover: true,
                draggable: true,
                progress: false,
                transition: Flip
            });
        }
    }
    return (
        <div>
            <ManageAdmins
                onSubmit={onSearch}
                searchUsername={searchUsername}
                setSearchUsername={setSearchUsername}
                searchUsernameError={searchUsernameError}
                currentAdmins={currentAdmins}

            />
        </div>
    );
}


export default ManageAdminsContainer;