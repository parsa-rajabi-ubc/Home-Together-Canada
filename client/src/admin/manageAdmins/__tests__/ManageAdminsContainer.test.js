/**
 * @Author:     Parsa Rajabi
 * @Created:    2021.3.12
 *
 * @Description: Manage Admins Test
 *
 */

import React from 'react';
import renderer from 'react-test-renderer'
import ManageAdminsContainer from "../ManageAdminsContainer";

describe('ManageAdminsContainer', () => {
    describe('Container test', () => {
        it('should match snapshot test regardless of the props', () => {
            //given
            const props = {
            }
            //when
            const component = renderer.create(<ManageAdminsContainer {...props}/>);
            const tree = component.toJSON();
            //then
            expect(tree).toMatchSnapshot();
        });
    });
});