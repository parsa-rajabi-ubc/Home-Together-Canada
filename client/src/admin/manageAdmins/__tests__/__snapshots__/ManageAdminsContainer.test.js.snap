// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageAdminsContainer Container test should match snapshot test regardless of the props 1`] = `
<div>
  <div>
    <h3
      className="account-summary-info-header"
    >
       
      Manage Admins
       
    </h3>
    <p
      className="account-summary-info-text"
    >
       
      Use the field below to search for members and provide them with admin privileges
       
    </p>
    <div
      className="p-4 flex mx-auto w-1/2 items-center"
    >
      <label>
        <label />
        <input
          className="undefined input m-auto mr-20 py-2"
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Search Member Username"
          type="text"
        />
      </label>
      <label>
        <input
          className="btn btn-green ml-10 w-2/3 py-2"
          disabled={false}
          onClick={[Function]}
          type="submit"
          value="Make Admin"
        />
      </label>
    </div>
    <label
      className="label text-lg mx-10"
    >
       
      Current Website Admins
       
    </label>
    <div
      className="mx-20 my-3"
    />
  </div>
</div>
`;
